File: AdminFrontend/services/AdvisorService.js
// @ts-check
/* global angular */
'use strict';

// Advisor Service
app.service('AdvisorService', ['ApiService', function(ApiService) {
    
    /**
     * Remove Vietnamese accents from text
     * @param {string} str - Input string
     * @returns {string} - String without accents
     */
    function removeVietnameseAccents(str) {
        if (!str) return str;
        
        // Normalize Unicode characters (decompose accented characters)
        return str.normalize('NFD')
            // Remove combining diacritical marks (accents)
            .replace(/[\u0300-\u036f]/g, '')
            // Replace đ/Đ with d/D
            .replace(/đ/g, 'd')
            .replace(/Đ/g, 'D');
    }
    
    /**
     * Get dashboard statistics
     * @param {object} filters - { facultyId, majorId, classId }
     * @param {boolean} useCache - Whether to use cache (default: true)
     */
    this.getDashboardStats = function(filters, useCache) {
        filters = filters || {};
        
        var queryParams = {
            facultyId: filters.facultyId || null,
            majorId: filters.majorId || null,
            classId: filters.classId || null
        };
        
        // Remove null values
        Object.keys(queryParams).forEach(function(key) {
            if (queryParams[key] === null || queryParams[key] === '') {
                delete queryParams[key];
            }
        });
        
        var cacheKey = 'advisor:dashboard:stats:' + JSON.stringify(queryParams);
        
        var options = {
            cache: useCache !== false,
            cacheKey: cacheKey,
            cacheTTL: 5 * 60 * 1000 // 5 minutes
        };
        
        return ApiService.get('/advisor/dashboard/stats', queryParams, options).then(function(response) {
            return response.data.data;
        });
    };
    
    /**
     * Get warning students (students with attendance or academic issues)
     * @param {object} params - { page, pageSize, facultyId, majorId, classId }
     * @param {boolean} useCache - Whether to use cache (default: false for pagination)
     */
    this.getWarningStudents = function(params, useCache) {
        params = params || {};
        
        var queryParams = {
            page: params.page || 1,
            pageSize: params.pageSize || 20,
            facultyId: params.facultyId || null,
            majorId: params.majorId || null,
            classId: params.classId || null
        };
        
        // Remove null values
        Object.keys(queryParams).forEach(function(key) {
            if (queryParams[key] === null || queryParams[key] === '') {
                delete queryParams[key];
            }
        });
        
        var options = {
            cache: useCache !== false && false, // Don't cache paginated results by default
            cacheTTL: 2 * 60 * 1000 // 2 minutes
        };
        
        return ApiService.get('/advisor/dashboard/warning-students', queryParams, options).then(function(response) {
            return response.data;
        });
    };
    
    /**
     * Get student detail with academic summary
     * @param {string} studentId
     * @param {boolean} useCache - Whether to use cache (default: true)
     */
    this.getStudentDetail = function(studentId, useCache) {
        if (!studentId) {
            return Promise.reject(new Error('Student ID is required'));
        }
        
        var cacheKey = 'advisor:student:detail:' + studentId;
        var options = {
            cache: useCache !== false,
            cacheKey: cacheKey,
            cacheTTL: 5 * 60 * 1000 // 5 minutes
        };
        
        return ApiService.get('/advisor/students/' + studentId, null, options).then(function(response) {
            return response.data.data;
        });
    };
    
    /**
     * Get student grades with filters (mandatory filter validation)
     * @param {string} studentId
     * @param {object} filters - { schoolYearId, semester, subjectId } - at least one required
     * @param {boolean} useCache - Whether to use cache (default: false)
     */
    this.getStudentGrades = function(studentId, filters, useCache) {
        if (!studentId) {
            return Promise.reject(new Error('Student ID is required'));
        }
        
        filters = filters || {};
        
        // Validate at least one filter
        if (!filters.schoolYearId && !filters.semester && !filters.subjectId) {
            return Promise.reject(new Error('Filter is required. Please provide at least one filter (schoolYearId, semester, or subjectId)'));
        }
        
        var queryParams = {
            schoolYearId: filters.schoolYearId || null,
            semester: filters.semester || null,
            subjectId: filters.subjectId || null
        };
        
        // Remove null values
        Object.keys(queryParams).forEach(function(key) {
            if (queryParams[key] === null || queryParams[key] === '') {
                delete queryParams[key];
            }
        });
        
        var options = {
            cache: useCache !== false && false, // Don't cache by default
            cacheTTL: 2 * 60 * 1000 // 2 minutes
        };
        
        return ApiService.get('/advisor/students/' + studentId + '/grades', queryParams, options).then(function(response) {
            return response.data;
        });
    };
    
    /**
     * Get student attendance with filters (mandatory filter validation)
     * @param {string} studentId
     * @param {object} filters - { schoolYearId, semester, subjectId } - at least one required
     * @param {boolean} useCache - Whether to use cache (default: false)
     */
    this.getStudentAttendance = function(studentId, filters, useCache) {
        if (!studentId) {
            return Promise.reject(new Error('Student ID is required'));
        }
        
        filters = filters || {};
        
        // Validate at least one filter
        if (!filters.schoolYearId && !filters.semester && !filters.subjectId) {
            return Promise.reject(new Error('Filter is required. Please provide at least one filter (schoolYearId, semester, or subjectId)'));
        }
        
        var queryParams = {
            schoolYearId: filters.schoolYearId || null,
            semester: filters.semester || null,
            subjectId: filters.subjectId || null
        };
        
        // Remove null values
        Object.keys(queryParams).forEach(function(key) {
            if (queryParams[key] === null || queryParams[key] === '') {
                delete queryParams[key];
            }
        });
        
        var options = {
            cache: useCache !== false && false, // Don't cache by default
            cacheTTL: 2 * 60 * 1000 // 2 minutes
        };
        
        return ApiService.get('/advisor/students/' + studentId + '/attendance', queryParams, options).then(function(response) {
            return response.data.data;
        });
    };
    
    /**
     * Get students list with filters (mandatory filter validation)
     * @param {object} params - { page, pageSize, filters } - filters must contain at least one filter
     * @param {boolean} useCache - Whether to use cache (default: false for pagination)
     */
    this.getStudents = function(params, useCache) {
        console.log('[AdvisorService] 🔍 getStudents() - Bắt đầu');
        console.log('[AdvisorService] 📥 Params nhận được:', params);
        
        params = params || {};
        
        // Validate at least one filter (unless showAll is true)
        var filters = params.filters || {};
        var showAll = params.showAll === true || filters.showAll === true;
        var hasFilter = false;
        
        if (filters.facultyId || filters.majorId || filters.classId || filters.cohortYear || 
            filters.search || filters.warningStatus || 
            filters.gpaMin !== undefined || filters.gpaMax !== undefined ||
            filters.attendanceRateMin !== undefined || filters.attendanceRateMax !== undefined) {
            hasFilter = true;
        }
        
        console.log('[AdvisorService] 📊 Validation:', {
            showAll: showAll,
            hasFilter: hasFilter,
            filters: filters
        });
        
        if (!hasFilter && !showAll) {
            console.error('[AdvisorService] ❌ Không có filter và showAll = false');
            return Promise.reject(new Error('Filter is required. Please provide at least one filter (faculty, major, class, cohort, search, warningStatus, gpa, or attendanceRate), or set showAll=true to view all students'));
        }
        
        var queryParams = {
            page: params.page || 1,
            pageSize: params.pageSize || 20,
            facultyId: filters.facultyId || null,
            majorId: filters.majorId || null,
            classId: filters.classId || null,
            cohortYear: filters.cohortYear || null,
            search: filters.search ? removeVietnameseAccents(filters.search) : null,
            warningStatus: filters.warningStatus || null,
            gpaMin: filters.gpaMin !== undefined ? filters.gpaMin : null,
            gpaMax: filters.gpaMax !== undefined ? filters.gpaMax : null,
            attendanceRateMin: filters.attendanceRateMin !== undefined ? filters.attendanceRateMin : null,
            attendanceRateMax: filters.attendanceRateMax !== undefined ? filters.attendanceRateMax : null,
            showAll: showAll ? true : null
        };
        
        // Remove null values
        Object.keys(queryParams).forEach(function(key) {
            if (queryParams[key] === null || queryParams[key] === '') {
                delete queryParams[key];
            }
        });
        
        console.log('[AdvisorService] 📤 Query params sau khi xử lý:', queryParams);
        console.log('[AdvisorService] 🌐 Gọi API: GET /advisor/students');
        
        var options = {
            cache: useCache !== false && false, // Don't cache paginated results by default
            cacheTTL: 2 * 60 * 1000 // 2 minutes
        };
        
        return ApiService.get('/advisor/students', queryParams, options).then(function(response) {
            console.log('[AdvisorService] ✅ API response:', response);
            console.log('[AdvisorService] 📊 Response data:', {
                hasData: !!response.data,
                dataType: typeof response.data,
                dataKeys: response.data ? Object.keys(response.data) : []
            });
            return response.data;
        }).catch(function(error) {
            console.error('[AdvisorService] ❌ API error:', error);
            console.error('[AdvisorService] ❌ Error details:', {
                message: error.message,
                data: error.data,
                status: error.status,
                statusText: error.statusText
            });
            throw error;
        });
    };
    
    /**
     * Get student GPA progress
     * @param {string} studentId - Student ID
     * @param {string} schoolYearId - Optional school year ID filter
     * @param {boolean} useCache - Whether to use cache
     */
    this.getStudentGpaProgress = function(studentId, schoolYearId, useCache) {
        if (!studentId) {
            return Promise.reject(new Error('Student ID is required'));
        }
        
        var queryParams = {};
        if (schoolYearId) {
            queryParams.schoolYearId = schoolYearId;
        }
        
        var cacheKey = 'advisor:student:gpa-progress:' + studentId + (schoolYearId ? ':' + schoolYearId : '');
        var options = {
            cache: useCache !== false,
            cacheKey: cacheKey,
            cacheTTL: 5 * 60 * 1000 // 5 minutes
        };
        
        return ApiService.get('/advisor/students/' + studentId + '/progress/gpa', queryParams, options).then(function(response) {
            return response.data.data;
        });
    };
    
    /**
     * Get student attendance progress
     * @param {string} studentId - Student ID
     * @param {string} schoolYearId - Optional school year ID filter
     * @param {boolean} useCache - Whether to use cache
     */
    this.getStudentAttendanceProgress = function(studentId, schoolYearId, useCache) {
        if (!studentId) {
            return Promise.reject(new Error('Student ID is required'));
        }
        
        var queryParams = {};
        if (schoolYearId) {
            queryParams.schoolYearId = schoolYearId;
        }
        
        var cacheKey = 'advisor:student:attendance-progress:' + studentId + (schoolYearId ? ':' + schoolYearId : '');
        var options = {
            cache: useCache !== false,
            cacheKey: cacheKey,
            cacheTTL: 5 * 60 * 1000 // 5 minutes
        };
        
        return ApiService.get('/advisor/students/' + studentId + '/progress/attendance', queryParams, options).then(function(response) {
            return response.data.data;
        });
    };
    
    /**
     * Get student trends and alerts
     * @param {string} studentId - Student ID
     * @param {boolean} useCache - Whether to use cache
     */
    this.getStudentTrends = function(studentId, useCache) {
        if (!studentId) {
            return Promise.reject(new Error('Student ID is required'));
        }
        
        var cacheKey = 'advisor:student:trends:' + studentId;
        var options = {
            cache: useCache !== false,
            cacheKey: cacheKey,
            cacheTTL: 5 * 60 * 1000 // 5 minutes
        };
        
        return ApiService.get('/advisor/students/' + studentId + '/progress/trends', null, options).then(function(response) {
            return response.data.data;
        });
    };
    
    /**
     * Get attendance warnings
     * @param {object} params - { page, pageSize, attendanceThreshold, filters }
     * @param {boolean} useCache - Whether to use cache
     */
    this.getAttendanceWarnings = function(params, useCache) {
        params = params || {};
        
        var queryParams = {
            page: params.page || 1,
            pageSize: params.pageSize || 100,
            attendanceThreshold: params.attendanceThreshold !== undefined ? params.attendanceThreshold : null,
            facultyId: params.filters?.facultyId || null,
            majorId: params.filters?.majorId || null,
            classId: params.filters?.classId || null,
            cohortYear: params.filters?.cohortYear || null
        };
        
        // Remove null values
        Object.keys(queryParams).forEach(function(key) {
            if (queryParams[key] === null || queryParams[key] === '') {
                delete queryParams[key];
            }
        });
        
        var options = {
            cache: useCache !== false && false, // Don't cache by default
            cacheTTL: 2 * 60 * 1000 // 2 minutes
        };
        
        return ApiService.get('/advisor/warnings/attendance', queryParams, options).then(function(response) {
            return response.data;
        });
    };
    
    /**
     * Get academic warnings
     * @param {object} params - { page, pageSize, gpaThreshold, filters }
     * @param {boolean} useCache - Whether to use cache
     */
    this.getAcademicWarnings = function(params, useCache) {
        params = params || {};
        
        var queryParams = {
            page: params.page || 1,
            pageSize: params.pageSize || 100,
            gpaThreshold: params.gpaThreshold !== undefined ? params.gpaThreshold : null,
            facultyId: params.filters?.facultyId || null,
            majorId: params.filters?.majorId || null,
            classId: params.filters?.classId || null,
            cohortYear: params.filters?.cohortYear || null
        };
        
        // Remove null values
        Object.keys(queryParams).forEach(function(key) {
            if (queryParams[key] === null || queryParams[key] === '') {
                delete queryParams[key];
            }
        });
        
        var options = {
            cache: useCache !== false && false, // Don't cache by default
            cacheTTL: 2 * 60 * 1000 // 2 minutes
        };
        
        return ApiService.get('/advisor/warnings/academic', queryParams, options).then(function(response) {
            return response.data;
        });
    };
    
    /**
     * Send warning email (single or bulk)
     * @param {object} params - { studentIds, warningType, customSubject, customMessage }
     */
    this.sendWarningEmail = function(params) {
        if (!params || !params.studentIds || params.studentIds.length === 0) {
            return Promise.reject(new Error('Student IDs are required'));
        }
        
        if (!params.warningType) {
            return Promise.reject(new Error('Warning type is required'));
        }
        
        var request = {
            studentIds: params.studentIds,
            warningType: params.warningType,
            customSubject: params.customSubject || null,
            customMessage: params.customMessage || null
        };
        
        return ApiService.post('/advisor/warnings/send-email', request).then(function(response) {
            return response.data;
        });
    };
    
    /**
     * Get warning configuration
     * @param {boolean} useCache - Whether to use cache
     */
    this.getWarningConfig = function(useCache) {
        var cacheKey = 'advisor:warning-config';
        var options = {
            cache: useCache !== false,
            cacheKey: cacheKey,
            cacheTTL: 10 * 60 * 1000 // 10 minutes
        };
        
        return ApiService.get('/advisor/warning-config', null, options).then(function(response) {
            return response.data.data;
        });
    };
    
    /**
     * Update warning configuration
     * @param {object} config - { attendanceThreshold, gpaThreshold, emailTemplate, emailSubject, autoSendEmails }
     */
    this.updateWarningConfig = function(config) {
        return ApiService.put('/advisor/warning-config', config).then(function(response) {
            return response.data;
        });
    };
}]);

File: EducationManagement/EducationManagement.BLL/Services/AdvisorService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using EducationManagement.Common.DTOs.Advisor;
using EducationManagement.DAL.Repositories;
using System.Linq;
using Microsoft.Extensions.Configuration;

namespace EducationManagement.BLL.Services
{
    public class AdvisorService
    {
        private readonly AdvisorRepository _advisorRepository;
        private readonly EmailService _emailService;
        private readonly IConfiguration _configuration;
        private readonly RetakeService? _retakeService;
        private readonly EnrollmentRepository? _enrollmentRepository;

        public AdvisorService(AdvisorRepository advisorRepository, EmailService emailService, 
            IConfiguration configuration, RetakeService? retakeService = null, 
            EnrollmentRepository? enrollmentRepository = null)
        {
            _advisorRepository = advisorRepository;
            _emailService = emailService;
            _configuration = configuration;
            _retakeService = retakeService;
            _enrollmentRepository = enrollmentRepository;
        }

        /// <summary>
        /// Get dashboard statistics with filters
        /// Advisor (Cố vấn phòng đào tạo) quản lý CHUNG TOÀN BỘ sinh viên của trường.
        /// Không filter theo advisor_id, chỉ filter theo Faculty, Major, Class (optional).
        /// Nếu không có filter, trả về thống kê toàn trường.
        /// </summary>
        public async Task<AdvisorDashboardStatsDto> GetDashboardStatsAsync(StudentFiltersDto? filters)
        {
            try
            {
                return await _advisorRepository.GetDashboardStatsAsync(
                    filters?.FacultyId,
                    filters?.MajorId,
                    filters?.ClassId
                );
            }
            catch (Exception ex)
            {
                throw new Exception($"Error getting dashboard stats: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Get warning students with filters
        /// Advisor có thể xem cảnh báo của TOÀN BỘ sinh viên (không filter theo advisor_id).
        /// Filter theo Faculty, Major, Class là optional để xem chi tiết.
        /// </summary>
        public async Task<(List<WarningStudentDto> Students, int TotalCount)> GetWarningStudentsAsync(
            int page = 1, int pageSize = 20, StudentFiltersDto? filters = null)
        {
            if (page < 1) page = 1;
            if (pageSize < 1) pageSize = 20;
            if (pageSize > 100) pageSize = 100;

            try
            {
                return await _advisorRepository.GetWarningStudentsAsync(
                    page, pageSize,
                    filters?.FacultyId,
                    filters?.MajorId,
                    filters?.ClassId
                );
            }
            catch (Exception ex)
            {
                throw new Exception($"Error getting warning students: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Get student detail with academic summary
        /// </summary>
        public async Task<AdvisorStudentDetailDto?> GetStudentDetailAsync(string studentId)
        {
            if (string.IsNullOrEmpty(studentId))
            {
                throw new ArgumentException("Student ID is required", nameof(studentId));
            }

            try
            {
                return await _advisorRepository.GetStudentDetailAsync(studentId);
            }
            catch (Exception ex)
            {
                throw new Exception($"Error getting student detail: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Get student grades with filters (mandatory filter validation)
        /// </summary>
        public async Task<(List<AdvisorStudentGradeDto> Grades, AdvisorStudentGradesSummaryDto Summary)> GetStudentGradesAsync(
            string studentId,
            string? schoolYearId = null,
            int? semester = null,
            string? subjectId = null)
        {
            if (string.IsNullOrEmpty(studentId))
            {
                throw new ArgumentException("Student ID is required", nameof(studentId));
            }

            // Mandatory filter validation: at least one filter must be provided
            if (string.IsNullOrEmpty(schoolYearId) && semester == null && string.IsNullOrEmpty(subjectId))
            {
                throw new ArgumentException(
                    "Filter is required. Please provide at least one filter (schoolYearId, semester, or subjectId)",
                    nameof(schoolYearId));
            }

            try
            {
                return await _advisorRepository.GetStudentGradesAsync(studentId, schoolYearId, semester, subjectId);
            }
            catch (Exception ex)
            {
                throw new Exception($"Error getting student grades: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Get student attendance with filters (mandatory filter validation)
        /// </summary>
        public async Task<AdvisorStudentAttendanceResponseDto> GetStudentAttendanceAsync(
            string studentId,
            string? schoolYearId = null,
            int? semester = null,
            string? subjectId = null)
        {
            if (string.IsNullOrEmpty(studentId))
            {
                throw new ArgumentException("Student ID is required", nameof(studentId));
            }

            // Mandatory filter validation: at least one filter must be provided
            if (string.IsNullOrEmpty(schoolYearId) && semester == null && string.IsNullOrEmpty(subjectId))
            {
                throw new ArgumentException(
                    "Filter is required. Please provide at least one filter (schoolYearId, semester, or subjectId)",
                    nameof(schoolYearId));
            }

            try
            {
                return await _advisorRepository.GetStudentAttendanceAsync(studentId, schoolYearId, semester, subjectId);
            }
            catch (Exception ex)
            {
                throw new Exception($"Error getting student attendance: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Get students list with filters (mandatory filter validation for performance)
        /// Advisor có thể xem TOÀN BỘ sinh viên, nhưng cần ít nhất 1 filter để tránh load quá nhiều data.
        /// Filter có thể là: Faculty, Major, Class, Cohort Year, Search, Warning Status, GPA range, Attendance rate.
        /// </summary>
        public async Task<(List<WarningStudentDto> Students, int TotalCount)> GetStudentsAsync(
            int page = 1,
            int pageSize = 20,
            StudentFiltersDto? filters = null)
        {
            if (page < 1) page = 1;
            if (pageSize < 1) pageSize = 20;
            if (pageSize > 100) pageSize = 100;

            // Mandatory filter validation: at least one filter must be provided (unless ShowAll is true)
            if (!filters?.ShowAll == true && 
                (filters == null ||
                 (string.IsNullOrEmpty(filters.FacultyId) &&
                  string.IsNullOrEmpty(filters.MajorId) &&
                  string.IsNullOrEmpty(filters.ClassId) &&
                  string.IsNullOrEmpty(filters.CohortYear) &&
                  string.IsNullOrEmpty(filters.Search) &&
                  string.IsNullOrEmpty(filters.WarningStatus) &&
                  !filters.GpaMin.HasValue &&
                  !filters.GpaMax.HasValue &&
                  !filters.AttendanceRateMin.HasValue &&
                  !filters.AttendanceRateMax.HasValue)))
            {
                throw new ArgumentException(
                    "Filter is required. Please provide at least one filter (faculty, major, class, cohort, search, warningStatus, gpa, or attendanceRate), or set showAll=true to view all students",
                    nameof(filters));
            }

            try
            {
                return await _advisorRepository.GetStudentsAsync(
                    page,
                    pageSize,
                    filters.Search,
                    filters.FacultyId,
                    filters.MajorId,
                    filters.ClassId,
                    filters.CohortYear,
                    filters.WarningStatus,
                    filters.GpaMin,
                    filters.GpaMax,
                    filters.AttendanceRateMin,
                    filters.AttendanceRateMax
                );
            }
            catch (Exception ex)
            {
                throw new Exception($"Error getting students: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Get student GPA progress by semester
        /// </summary>
        public async Task<StudentGpaProgressDto?> GetStudentGpaProgressAsync(string studentId, string? schoolYearId = null)
        {
            if (string.IsNullOrEmpty(studentId))
            {
                throw new ArgumentException("Student ID is required", nameof(studentId));
            }

            try
            {
                return await _advisorRepository.GetStudentGpaProgressAsync(studentId, schoolYearId);
            }
            catch (Exception ex)
            {
                throw new Exception($"Error getting student GPA progress: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Get student attendance progress by semester
        /// </summary>
        public async Task<StudentAttendanceProgressDto?> GetStudentAttendanceProgressAsync(string studentId, string? schoolYearId = null)
        {
            if (string.IsNullOrEmpty(studentId))
            {
                throw new ArgumentException("Student ID is required", nameof(studentId));
            }

            try
            {
                return await _advisorRepository.GetStudentAttendanceProgressAsync(studentId, schoolYearId);
            }
            catch (Exception ex)
            {
                throw new Exception($"Error getting student attendance progress: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Get student trends and alerts
        /// </summary>
        public async Task<StudentTrendsDto?> GetStudentTrendsAsync(string studentId)
        {
            if (string.IsNullOrEmpty(studentId))
            {
                throw new ArgumentException("Student ID is required", nameof(studentId));
            }

            try
            {
                return await _advisorRepository.GetStudentTrendsAsync(studentId);
            }
            catch (Exception ex)
            {
                throw new Exception($"Error getting student trends: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Get attendance warnings
        /// </summary>
        public async Task<(List<AttendanceWarningDto> Warnings, int TotalCount)> GetAttendanceWarningsAsync(
            int page = 1,
            int pageSize = 100,
            decimal? attendanceThreshold = null,
            StudentFiltersDto? filters = null)
        {
            if (page < 1) page = 1;
            if (pageSize < 1) pageSize = 20;
            if (pageSize > 100) pageSize = 100;

            // Get threshold from parameter, database config, or default
            var threshold = attendanceThreshold ?? await GetAttendanceThresholdAsync();

            try
            {
                return await _advisorRepository.GetAttendanceWarningsAsync(
                    page,
                    pageSize,
                    threshold,
                    filters?.FacultyId,
                    filters?.MajorId,
                    filters?.ClassId,
                    filters?.CohortYear
                );
            }
            catch (Exception ex)
            {
                throw new Exception($"Error getting attendance warnings: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Get academic warnings
        /// </summary>
        public async Task<(List<AcademicWarningDto> Warnings, int TotalCount)> GetAcademicWarningsAsync(
            int page = 1,
            int pageSize = 100,
            decimal? gpaThreshold = null,
            StudentFiltersDto? filters = null)
        {
            if (page < 1) page = 1;
            if (pageSize < 1) pageSize = 20;
            if (pageSize > 100) pageSize = 100;

            // Get threshold from parameter, database config, or default
            var threshold = gpaThreshold ?? await GetGpaThresholdAsync();

            try
            {
                return await _advisorRepository.GetAcademicWarningsAsync(
                    page,
                    pageSize,
                    threshold,
                    filters?.FacultyId,
                    filters?.MajorId,
                    filters?.ClassId,
                    filters?.CohortYear
                );
            }
            catch (Exception ex)
            {
                throw new Exception($"Error getting academic warnings: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Send warning email to a student
        /// </summary>
        public async Task SendWarningEmailAsync(string studentId, string warningType, string? customSubject = null, string? customMessage = null)
        {
            if (string.IsNullOrEmpty(studentId))
            {
                throw new ArgumentException("Student ID is required", nameof(studentId));
            }

            if (string.IsNullOrEmpty(warningType))
            {
                throw new ArgumentException("Warning type is required", nameof(warningType));
            }

            try
            {
                var studentInfo = await _advisorRepository.GetStudentInfoForEmailAsync(studentId);
                if (studentInfo == null)
                {
                    throw new Exception("Student not found");
                }

                var (email, fullName, className, gpa, absenceRate) = studentInfo.Value;

                if (string.IsNullOrEmpty(email))
                {
                    throw new Exception("Student email not found");
                }

                className ??= "N/A";

                switch (warningType.ToLower())
                {
                    case "attendance":
                        if (!absenceRate.HasValue)
                        {
                            throw new Exception("Absence rate not available for student");
                        }
                        await _emailService.SendAttendanceWarningEmailAsync(
                            email, fullName, className, absenceRate.Value, customSubject, customMessage);
                        break;

                    case "academic":
                        if (!gpa.HasValue)
                        {
                            throw new Exception("GPA not available for student");
                        }
                        await _emailService.SendAcademicWarningAsync(
                            email, fullName, className, gpa.Value, customSubject, customMessage);
                        break;

                    case "both":
                        if (!gpa.HasValue || !absenceRate.HasValue)
                        {
                            throw new Exception("GPA or absence rate not available for student");
                        }
                        await _emailService.SendBothWarningEmailAsync(
                            email, fullName, className, gpa.Value, absenceRate.Value, customSubject, customMessage);
                        break;

                    default:
                        throw new ArgumentException($"Invalid warning type: {warningType}", nameof(warningType));
                }
            }
            catch (Exception ex)
            {
                throw new Exception($"Error sending warning email: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Send bulk warning emails
        /// </summary>
        public async Task SendBulkWarningEmailAsync(List<string> studentIds, string warningType, string? customSubject = null, string? customMessage = null)
        {
            if (studentIds == null || studentIds.Count == 0)
            {
                throw new ArgumentException("Student IDs are required", nameof(studentIds));
            }

            if (string.IsNullOrEmpty(warningType))
            {
                throw new ArgumentException("Warning type is required", nameof(warningType));
            }

            var errors = new List<string>();

            foreach (var studentId in studentIds)
            {
                try
                {
                    await SendWarningEmailAsync(studentId, warningType, customSubject, customMessage);
                }
                catch (Exception ex)
                {
                    errors.Add($"Student {studentId}: {ex.Message}");
                }
            }

            if (errors.Count > 0)
            {
                throw new Exception($"Some emails failed to send: {string.Join("; ", errors)}");
            }
        }

        /// <summary>
        /// Get warning configuration from database with fallback to appsettings.json
        /// </summary>
        public async Task<WarningConfigDto> GetWarningConfigAsync()
        {
            try
            {
                // Try to get config from database first
                var dbConfig = await _advisorRepository.GetWarningConfigAsync();
                
                if (dbConfig != null)
                {
                    return dbConfig; // Return config from database
                }
            }
            catch (Exception)
            {
                // If database read fails, fall back to appsettings.json
            }

            // Fallback to appsettings.json if database config doesn't exist
            return new WarningConfigDto
            {
                AttendanceThreshold = decimal.Parse(_configuration["Advisor:WarningThresholds:Attendance"] ?? "20.0"),
                GpaThreshold = decimal.Parse(_configuration["Advisor:WarningThresholds:Gpa"] ?? "2.0"),
                EmailTemplate = _configuration["Advisor:Email:Template"],
                EmailSubject = _configuration["Advisor:Email:Subject"],
                AutoSendEmails = bool.Parse(_configuration["Advisor:Email:AutoSend"] ?? "false")
            };
        }

        /// <summary>
        /// Helper method to get attendance threshold from database or config file
        /// </summary>
        private async Task<decimal> GetAttendanceThresholdAsync()
        {
            try
            {
                var config = await GetWarningConfigAsync();
                return config.AttendanceThreshold;
            }
            catch (Exception)
            {
                return decimal.Parse(_configuration["Advisor:WarningThresholds:Attendance"] ?? "20.0");
            }
        }

        /// <summary>
        /// Helper method to get GPA threshold from database or config file
        /// </summary>
        private async Task<decimal> GetGpaThresholdAsync()
        {
            try
            {
                var config = await GetWarningConfigAsync();
                return config.GpaThreshold;
            }
            catch (Exception)
            {
                return decimal.Parse(_configuration["Advisor:WarningThresholds:Gpa"] ?? "2.0");
            }
        }

        /// <summary>
        /// Update warning configuration (saved to database)
        /// </summary>
        public async Task UpdateWarningConfigAsync(WarningConfigDto config, string? updatedBy = null)
        {
            // Validate configuration
            if (config.AttendanceThreshold < 0 || config.AttendanceThreshold > 100)
            {
                throw new ArgumentException("Ngưỡng cảnh báo điểm danh phải nằm trong khoảng 0-100%", nameof(config));
            }

            if (config.GpaThreshold < 0 || config.GpaThreshold > 10)
            {
                throw new ArgumentException("Ngưỡng cảnh báo GPA phải nằm trong khoảng 0-10", nameof(config));
            }

            try
            {
                // Save to database
                await _advisorRepository.UpdateWarningConfigAsync(config, updatedBy);
            }
            catch (Exception ex)
            {
                throw new Exception($"Lỗi khi cập nhật cấu hình: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Check and send warning after attendance is marked (Event-Driven)
        /// Called automatically when attendance is created
        /// </summary>
        public async Task CheckAndSendWarningAfterAttendance(string studentId, string classId)
        {
            if (string.IsNullOrWhiteSpace(studentId))
                return;

            if (string.IsNullOrWhiteSpace(classId))
                return;

            try
            {
                // 1. Get absence rate for this student in this class
                var absenceRate = await _advisorRepository.GetStudentAbsenceRateByClassAsync(studentId, classId);
                
                if (!absenceRate.HasValue)
                    return; // No attendance data yet

                // 2. Get threshold from config (database with fallback to appsettings.json)
                var threshold = await GetAttendanceThresholdAsync();
                var minDaysBetweenWarnings = int.Parse(_configuration["Advisor:WarningSettings:MinDaysBetweenWarnings"] ?? "7");
                
                // Check if auto-send is enabled
                var config = await GetWarningConfigAsync();
                if (!config.AutoSendEmails)
                {
                    return; // Auto-send is disabled, don't send warning
                }

                // 3. Check if exceeds threshold
                if (absenceRate.Value > threshold)
                {
                    // 4. Check last warning sent (avoid spam)
                    var lastWarning = await _advisorRepository.GetLastWarningSentAsync(studentId);
                    var shouldSend = lastWarning == null || 
                                    lastWarning.Value < DateTime.Now.AddDays(-minDaysBetweenWarnings);

                    if (shouldSend)
                    {
                        // 5. Get student info
                        var studentInfo = await _advisorRepository.GetStudentInfoForEmailAsync(studentId);
                        if (studentInfo.HasValue)
                        {
                            var (email, fullName, className, gpa, _) = studentInfo.Value;

                            if (!string.IsNullOrEmpty(email))
                            {
                                // 6. Send warning email
                                await _emailService.SendAttendanceWarningEmailAsync(
                                    email,
                                    fullName,
                                    className ?? "N/A",
                                    absenceRate.Value
                                );

                                // 7. Update last warning sent
                                await _advisorRepository.UpdateLastWarningSentAsync(studentId);

                                // 8. ✅ Auto-create retake record if absence rate > threshold
                                if (_retakeService != null && _enrollmentRepository != null)
                                {
                                    try
                                    {
                                        // Get enrollment_id from studentId and classId
                                        var enrollments = await _enrollmentRepository.GetByStudentIdAsync(studentId);
                                        var enrollment = enrollments.FirstOrDefault(e => e.ClassId == classId && 
                                            e.EnrollmentStatus == "APPROVED" && e.DeletedAt == null);
                                        
                                        if (enrollment != null)
                                        {
                                            // Trigger retake check (async, don't wait)
                                            _ = Task.Run(async () =>
                                            {
                                                try
                                                {
                                                    await _retakeService.CheckAndCreateRetakeAsync(enrollment.EnrollmentId);
                                                }
                                                catch (Exception ex2)
                                                {
                                                }
                                            });
                                        }
                                    }
                                    catch (Exception ex2)
                                    {
                                    }
                                }
                            }
                        }
                    }
                }
            }
            catch (Exception ex)
            {
            }
        }
    }
}

